signer/src/block_notifier.rs

```diff
@@ -0,0 +1,177 @@
+        let retry_interval = Duration::from_secs(config.retry_interval);
+        let max_retry_attempts = config.max_retry_attempts;
+        let ping_interval = Duration::from_secs(config.ping_interval);
+        let subscribe_interval = Duration::from_secs(config.subscribe_interval);
+        Ok(ElectrumBlockNotifier {
+            client,
+            retry_interval,
+            max_retry_attempts,
+            ping_interval,
+            subscribe_interval,
+        })
+    }
+
+    /// Creates a new instance of `ElectrumBlockNotifier` from system config.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ElectrumBlockNotifier`.
+    pub fn new() -> Result<Self, ElectrumError> {
+        let config = &SETTINGS.block_notifier;
+        Self::from_config(config.clone())
+    }
+
+    /// The notify loop that handles block header notifications and sends them to the broadcast channel.
+    async fn notify_loop(
+        client: Arc<Mutex<Client>>,
+        retry_interval: Duration,
+        max_retry_attempts: u32,
+        ping_interval: Duration,
+        subscribe_interval: Duration,
+        sender: Sender<Result<BlockHash, Error>>,
+    ) {
+        let mut retry_attempts = 0;
+        let mut ping_interval = interval(ping_interval);
+        let mut subscribe_interval = interval(subscribe_interval);
+        let mut continue_loop = true;
+
+        debug!("Beginning to poll for block headers");
+
+        while continue_loop {
+            tokio::select! {
+                // Concurrent keep-alive check
+                _ = ping_interval.tick() => {
+                    if client.lock().await.ping().is_err() {
+                        error!("Ping failed, attempting to reconnect...");
+                        retry_attempts += 1;
+                        if retry_attempts >= max_retry_attempts {
+                            error!("Max retry attempts reached, giving up.");
+                            continue_loop = false;
+                        }
+                    } else {
+                        retry_attempts = 0; // Reset retry attempts on successful ping
+                    }
+                }
+                // Periodic block header subscription
+                _ = subscribe_interval.tick() => {
+                    let client_clone = Arc::clone(&client);
+                    let sender_clone = sender.clone();
+
+                    task::spawn_blocking(move || {
+                        match client_clone.blocking_lock().block_headers_subscribe() {
+                            Ok(HeaderNotification { header, .. }) => {
+                                let block_hash = header.block_hash();
+                                let _ = sender_clone.send(Ok(block_hash));
+                            }
+                            Err(e) => {
+                                let _ = sender_clone.send(Err(Error::Electrum(Arc::new(e))));
```

Comment:
Nit: I don't think the electrum errors should be propagated to the subscribers. Instead we should either have some sort of retry logic and only send block headers to the subscribers when the client succeeds.

If you don't have time to implement this, feel free to merge as-is and we can take that as a follow-up.